_posts/2016-05-19-big-data-cap-theorem-plain-english.md

```diff
@@ -0,0 +1,111 @@
+
+This article will pick some ideas from those conversations and will try to explain, by means of a metaphor, the implications of BigData and the CAP theorem[^other].
+
+## Scaling vertically vs horizontally
+
+Imagine that you own a [library](https://en.wikipedia.org/wiki/Library) (yes, that will be our metaphor). You have lots of books and magazines in it but, at some point, you don't have enough space in the shelves, so you decide to buy some more.
+
+<img alt="Adding floors" style="float: right;" src="/assets/images/big-data-cap-theorem/scafolding.jpg">
+
+After some weeks, you've filled the five floors of your library with shelves but you still don't have enough space (disk) for all the books you want to keep. This is when you decide you need to add floors to the library building.
+
+With more books you also need to grow your library index, so you need to buy more cards (memory). You'll probably need to hire more librarians (cpu). Also, there is a point there are queues outside the building (too many visitors) so you need to open more entrance doors (network).
+
+As you have imagined, this is what it's called "scaling up" or "scaling vertically"[^scale]. You find your bottleneck (e.g. memory) and just make it bigger.
+
+Someday you'll see that the cost of expanding your library is prohibitive or, directly that it is physically impossible to add more shelves, indexes or doors. That day is the day you realise you need to build another library.
+
+And yes, your guess is correct, creating a library network and adding more nodes (buildings) to it would be to "scale out" or "scale horizontally" your library.
+
+## Sharding, redundancy and replication factor
+
+### Sharding
+Now that you have multiple buildings you need to take some important decisions. The first one is to decide how to distribute your books among the different libraries (sharding[^shard]).
+
+One possible and easy solution would be to distribute them by author. Books of authors whose surname starts with A, B, C or D go to building number 1. Authors with surname starting with E, F, G or H go to building number 2 and so on.
+
+You soon realise a problem with that approach. The authors surnames are not distributed equally and some authors (like Terry Pratchet) have written many more books than others. The implication is that some of your buildings are half empty while some others are too busy.
+
+You need to choose a better criterion (partition or shard key) that is easy to compute and well distributed. There are several options and there is not a good answer that will work in any scenario (it will depend on how your index is created, what queries can you do, what are the most common queries people is doing...). But, as an example, hashing the book title and using the modulus to decide which building the book is in would provide a better distribution of the books.
```

Comment:
A good "proper" english rule of thumb is to try to never start a sentence with "but". Most of the time, you can swap it for "However":

> However, as an example, hashing the book title and using the modulus to decide which building the book is in would provide a better distribution of the books.